Python Model API package: add main documentation (#3268)
openvinotoolkit/open_model_zoo: demos/common/python/openvino/model_zoo/model_api/README.md

# Python* Model API package

Model API package is a set of wrapper classes for particular tasks and model architectures, simplifying data preprocess and postprocess as well as routine procedures (model loading, asynchronous execution, etc...)
An application feeds model class with input data, then the model returns postprocessed output data in user-friendly format.

## Package structure

> The Python* Model API consists of 3 libraries:

Reviewer: ```suggestion
The Model API consists of 3 libraries:
```
Python mentions are redundant in paragraphs texts, can't we leave them just in headers? At this point it's already obvious that it is python package.

Author: Removed in paragraphs texts

> * _adapters_ implements a common interface to allow Model API wrappers usage with different executors: OpenVINO, OVMS. See [Model API adapters](#model-api-adapters) section

Reviewer: ```suggestion
* _adapters_ implements a common interface to allow Model API wrappers usage with different executors. See [Model API adapters](#model-api-adapters) section
```
There shouldn't be examples. We will definitely forget to update here when more adapters will be introduced. Let's leave it only to the adapters section.

`OpenvinoAdapter` hides the OpenVINO™ toolkit API, which allows Model API wrappers launching with models represented in Intermediate Representation (IR) format.
It accepts a path to either `xml` model file or `onnx` model file.

#### OpenVINO Model Server Adapter

`OVMSAdapter` hides the OpenVINO Model Server python client API, which allows Model API wrappers launching with models served by OVMS.

Refer to __[`OVMSAdapter`](adapters/ovms_adapter.md)__ to learn about running demos with OVMS.

For using OpenVINO Model Server Adapter you need to install the package with extra module:
```sh
pip install <omz_dir>/demos/common/python[ovms]
```

## Model API Pipelines

Model API Pipelines represent the high-level wrappers upon the input data and accessing model results management.
They perform the data submission for model inference, verification of inference status, whether the result is ready or not, and results accessing.

The `AsyncPipeline` is available, which handles the asynchronous execution of a single model.

## Ready-to-use Model API solutions

To apply Model API wrappers in custom applications, learn the provided example of common scenario of how to use Python* Model API.

>  In the example, the SSD architecture is used to predict bounding boxes on input image `"sample.png"`. The model execution is produced by `OpenvinoAdapter`, therefore we submit the path to the model's `xml` file.

Reviewer: @anzhella-pankratova @Wovchena @vladimir-dudnik There could be a confusion with "architecture" definition, especially in documentation. Some people already asked me, why rcnn models in object_detection_demo relate to ssd architecture type. If there any suggestions how to avoid this misunderstandings? Maybe we should rename architecture_type flag in demos and not mention architecture at all, because principle on what we unify models in one wrapper is entirely about task and inputs/outputs structure similarity.

Reviewer 2: Although we refuse to guarantee custom model works, but it should be flexible enough to allow run model which is not from OMZ, but compatible by topology, inputs/outputs with our. The last example, is `yoloface` model, this can work with OMZ demo when properly converted to IR, and that would be excessive to require customer duplicate our yolo wrapper just because model name is different then OMZ model (or even was not set at conversion time).

Reviewer: @vladimir-dudnik so what do we decide, for now leave as is?

Reviewer 3: I would answer yes. Imagine you never asked :)

* _models_ - implement wrappers for each architecture. See [Model API Wrappers](#model-api-wrappers) section
* _pipelines_ - implement pipelines for model inference and manage the synchronous/asynchronous execution. See [Model API Pipelines](#model-api-pipelines) section

## Building Python* Model API package
For installation Python (version 3.6 or higher) is required. The installation is available from source.

Use the following command to install Python* Model API from source:
```sh
pip install <omz_dir>/demos/common/python
```

Alternatively, you can generate the package using a wheel. Follow the steps below:
1. Build the wheel.

```sh
python <omz_dir>/demos/common/python/setup.py bdist_wheel
```
The wheel should appear in the dist folder.
Name example: `openmodelzoo_modelapi-0.0.0-py3-none-any.whl`

2. Install the package in the clean environment with `--force-reinstall` key.
```sh
pip install openmodelzoo_modelapi-0.0.0-py3-none-any.whl --force-reinstall
```

To verify the package is installed, you might use the following command:
```sh
python -c "from openvino.model_zoo import model_api"
```

> > **NOTE**: On Linux and macOS, you may need to type `python3` instead of `python`. You may also need to [install pip](https://pip.pypa.io/en/stable/installation/).

Reviewer 4: probably it is extra info, it should be obvious that user should use appropriate python version, anyway it is not cover all cases (e.g. working in virual environment does not requered to use python3 / pip3 tools if you do not create environment with python2

Author: Removed the note

> For example, on Ubuntu execute the following command to get pip installed: `sudo apt install python3-pip`.

## Model API Wrappers

The Python* Model API package suggests ready-to-use model wrappers, which implement standardized preprocessing/postprocessing functions per "task type" and might be reused in applications as "black-box" models.
Also, the simple wrapper interface allows the creation of custom wrappers covering different architectures.

The following tasks can be solved with wrappers usage:

| Task type                  | Model API wrappers |
|----------------------------|--------------------|
| Background Matting         | `VideoBackgroundMatting`, `ImageMattingWithBackground` |
| Classification             | `Classification` |
| Deblurring                 | `Deblurring` |
| Human Pose Estimation      | `HpeAssociativeEmbedding`, `OpenPose` |
| Instance Segmentation      | `MaskRCNNModel`, `YolactModel` |
| Monocular Depth Estimation | `MonoDepthModel` |
| Named Entity Recognition   | `BertNamedEntityRecognition` |

> |  Object Detection          | `CenterNet`, `DETR`, `CTPN`, `FaceBoxes`, `RetinaFace`, `RetinaFacePyTorch`, `SSD`, `UltraLightweightFaceDetection`, `YOLO`, `YoloV3ONNX`, `YoloV4`, `YOLOF`, `YOLOX` |

Reviewer 4: can bullet points will be used inside the table for multiple wrappers line? it will be more accurate, then just comma separation, when you have a lot of wrappers like for object detection

Author: Added bullet points using html lists

| Question Answering         | `BertQuestionAnswering` |
| Salient Object Detection   | `SalientObjectDetectionModel` |
| Semantic Segmentation      | `SegmentationModel` |

## Model API Adapters

Model API wrappers are executor-agnostic, meaning it does not implement the specific model inference or model loading, instead it can be used with different executors having the implementation of common interface methods in adapter class respectively.

Currently, `OpenvinoAdapter` and `OVMSAdapter` are supported.

> #### OpenVINO executor

Reviewer 4: probably better name them adapters, looks misleading if you do not provide what is executor in your terms, but call class of adapter

Author: Renamed

Once the SSD model wrapper instance is created, we get the predictions by the model in one line: `ssd_model(input_data)` - the wrapper performs the preprocess method, synchronous inference on OpenVINO side and postprocess method.

```python
import cv2
from openvino.model_zoo.model_api.models import SSD
from openvino.model_zoo.model_api.adapters import OpenvinoAdapter, create_core

# a helper function for bboxes visualization
def draw_detections(image, detections):
    for detection in detections:
        class_id = int(detection.id)
        color = (255, 0, 0)
        det_label = '#{}'.format(class_id)
        xmin, ymin, xmax, ymax = detection.get_coords()
        cv2.rectangle(image, (xmin, ymin), (xmax, ymax), color, 2)
        cv2.putText(image, '{} {:.1%}'.format(det_label, detection.score),
                    (xmin, ymin - 7), cv2.FONT_HERSHEY_COMPLEX, 0.6, color, 1)
    return image

def main():
    input_data = cv2.imread("sample.png")
    model_path = "public/mobilenet-ssd/FP32/mobilenet-ssd.xml"

    model_adapter = OpenvinoAdapter(create_core(), model_path, device="CPU")
    ssd_model = SSD(model_adapter, preload=True)

>     results = ssd_model(input_data)

Reviewer 4: I think these lines for demonstration how api should be used is enough, also you can say that you can find a lot of examples in demos. It will be great if we can have list of model api supported demos also

Author: Added the list of such demos.

> I think these lines for demonstration how api should be used is enough, also you can say that you can find a lot of examples in demos  

The reference to OMZ demos is good, but I found useful also to add the visualization to this snippet. Someone can just take this code, launch and see clearly how the Model API works. With only these lines it's not obvious, what the format of `results` is (as we remember from demos, it is the list of `Detection` structure instances).

So, I think this example for sync inference should be self-contained enough and representative as possible. For the more complex usages, like async inference, there is the recommendation to refer to OMZ demos

Reviewer 4: I confused by main function and other stuff which make example more similar to ready application, my opinion that we should just provide api with comments which line used for what (where model loaded, where data provided, where results should be visualized and so on) without wrapping to application

Reviewer 4: BTW, what about submitting jupiter notebook to https://github.com/openvinotoolkit/openvino_notebooks
with demonstration how to work with model api? Is it planned for future?

Author: Updated example according to your suggestions.

The `openvino_notebooks` team waits then the package becomes the part of openvino-dev package as I remember. Possibly the notebook with model API with demonstration will be useful, we should discuss it with notebooks maintainers.

# Python* Model API package

Model API package is a set of wrapper classes for particular tasks and model architectures, simplifying data preprocess and postprocess as well as routine procedures (model loading, asynchronous execution, etc...)

> Model API wrappers hide the specific code inside and work as black-box: application feeds model class with input data, then the model returns post-processed output data in user-friendly format.

Reviewer 3: ```suggestion
An application feeds model class with input data, then the model returns post-processed output data in user-friendly format.
```
It doesn't bring any information

> Model API wrappers hide the specific code inside and work as black-box: application feeds model class with input data, then the model returns post-processed output data in user-friendly format.

Reviewer 3: One line above you have `postprocess` without `-`, but here is `post-processed`

## Package structure

The Python* Model API consists of 3 libraries:

> * _adapters_ - implement a common interface to allow Model API wrappers usage with different executors: OpenVINO, ONNX, etc. See [Model API adapters](#model-api-adapters) section
> * _models_ - implement wrappers for each architecture. See [Model API Wrappers](#model-api-wrappers) section
> * _pipelines_ - implement pipelines for model inference and manage the synchronous/asynchronous execution. See [Model API Pipelines](#model-api-pipelines) section

Reviewer 3: ```suggestion
* _adapters_ implements a common interface to allow Model API wrappers usage with different executors: OpenVINO, OVMS. See [Model API adapters](#model-api-adapters) section
* _models_ implements wrappers for each architecture. See [Model API Wrappers](#model-api-wrappers) section
* _pipelines_ implements pipelines for model inference and manage the synchronous/asynchronous execution. See [Model API Pipelines](#model-api-pipelines) section
```
Why ` - `? These are regular sentences

### Prerequisites

The package requires
- Python (version 3.6 or higher)

> - OpenVINO™ toolkit

Reviewer 3: OpenCV is missing

Author: OpenCV is installed from requirements automatically

If you build Python* Model API package from source, you should install the OpenVINO™ toolkit. See the options:

Use installation package for [Intel® Distribution of OpenVINO™ toolkit](https://www.intel.com/content/www/us/en/developer/tools/openvino-toolkit-download.html) or build the open-source version available in the [OpenVINO GitHub repository](https://github.com/openvinotoolkit/openvino) using the [build instructions](https://github.com/openvinotoolkit/openvino/wiki/BuildingCode).

Also, you can install the OpenVINO Python\* package via the command:
 ```sh
pip install openvino
 ```

> ## Building Python* Model API package

Reviewer 3: ```suggestion
## Installing Python* Model API package
```
The point is to get it installed, not to generate whl for redistribution

Use the following command to install Python* Model API from source:
```sh
pip install <omz_dir>/demos/common/python
```

Alternatively, you can generate the package using a wheel. Follow the steps below:
1. Build the wheel.

```sh
python <omz_dir>/demos/common/python/setup.py bdist_wheel
```
The wheel should appear in the dist folder.
Name example: `openmodelzoo_modelapi-0.0.0-py3-none-any.whl`

2. Install the package in the clean environment with `--force-reinstall` key.
```sh
pip install openmodelzoo_modelapi-0.0.0-py3-none-any.whl --force-reinstall
```

To verify the package is installed, you might use the following command:
```sh
python -c "from openvino.model_zoo import model_api"
```

## Model API Wrappers

> The Python* Model API package suggests ready-to-use model wrappers, which implement standardized preprocessing/postprocessing functions per "task type" and might be reused in applications as "black-box" models.

Reviewer 4: ```suggestion
The Python* Model API package suggests ready-to-use model wrappers, which implement standardized preprocessing/postprocessing functions per "task type" and  incapsulate model-specific logic, which allows usage of different models in a unified manner inside the application
```

> The Python* Model API package suggests ready-to-use model wrappers, which implement standardized preprocessing/postprocessing functions per "task type" and incapsulate model-specific logic for usage of different models in a unified manner inside the application.

Reviewer 3: ```suggestion
The Python* Model API package provides model wrappers, which implement standardized preprocessing/postprocessing functions per "task type" and incapsulate model-specific logic for usage of different models in a unified manner inside the application.
```
Every class is ready-to-use

> Also, the simple wrapper interface allows the creation of custom wrappers covering different architectures.

Reviewer 4: ```suggestion
The wrapper interface is simple and flexible, which gives capabilities for the creation of custom wrappers covering different architectures.
```

> The wrapper interface is simple and flexible, which gives capabilities for the creation of custom wrappers covering different architectures and use cases.
> 

Reviewer 3: ```suggestion
```
It doesn't bring anything

Author: Why? Shouldn't we mention that one might create custom wrapper class according to the flexible interface? @vladimir-dudnik

Reviewer 3: That's Python. Anyone can do anything with your classes

Reviewer 2: Did not get your point. Anyone can do anything, including "creation of custom wrappers covering different architectures and use cases." Then statement is not wrong?

Reviewer 3: The statement is not wrong, but
> It doesn't bring anything

Reviewer 2: ok, let's remove this statement

| Question Answering         |  <ul><li>`BertQuestionAnswering`</li></ul> |
| Salient Object Detection   |  <ul><li>`SalientObjectDetectionModel`</li></ul> |
| Semantic Segmentation      |  <ul><li>`SegmentationModel`</li></ul> |

## Model API Adapters

Model API wrappers are executor-agnostic, meaning it does not implement the specific model inference or model loading, instead it can be used with different executors having the implementation of common interface methods in adapter class respectively.

Currently, `OpenvinoAdapter` and `OVMSAdapter` are supported.

#### OpenVINO Adapter

`OpenvinoAdapter` hides the OpenVINO™ toolkit API, which allows Model API wrappers launching with models represented in Intermediate Representation (IR) format.
It accepts a path to either `xml` model file or `onnx` model file.

#### OpenVINO Model Server Adapter

`OVMSAdapter` hides the OpenVINO Model Server python client API, which allows Model API wrappers launching with models served by OVMS.

Refer to __[`OVMSAdapter`](adapters/ovms_adapter.md)__ to learn about running demos with OVMS.

For OpenVINO Model Server Adapter employment, you need to install the requirements:
```sh

> pip install <omz_dir>/demos/common/python/requirements_ovms.txt

Reviewer 4: can it be done with extra modules in model api install?

Reviewer 4: -r missed in command

Author: > can it be done with extra modules in model api install?

I'll investigate it

pip install <omz_dir>/demos/common/python[ovms]
```

## Model API Pipelines

Model API Pipelines represent the high-level wrappers upon the input data and accessing model results management.

> It performs the data submission for model inference, verification of inference status, whether the result is ready or not, and results accessing.

Reviewer 3: ```suggestion
They perform the data submission for model inference, verification of inference status, whether the result is ready or not, and results accessing.
```

The `AsyncPipeline` is available, which handles the asynchronous execution of a single model.

## Ready-to-use Model API solutions

To apply Model API wrappers in custom applications, learn the provided example of common scenario of how to use Python* Model API.

 In the example, the SSD architecture is used to predict bounding boxes on input image `"sample.png"`. The model execution is produced by `OpenvinoAdapter`, therefore we submit the path to the model's `xml` file. The model is loaded on a CPU device inside the adapter.

Once the SSD model wrapper instance is created, we get the predictions by the model in one line: `ssd_model(input_data)` - the wrapper performs the preprocess method, synchronous inference on OpenVINO™ toolkit side and postprocess method.

```python
import cv2

> from openvino.model_zoo.model_api.models import SSD

Reviewer 4: ```suggestion
# import model wrapper class
from openvino.model_zoo.model_api.models import SSD
```

> from openvino.model_zoo.model_api.adapters import OpenvinoAdapter, create_core

Reviewer 4: ```suggestion
# import inference adapter and helper for runtime setup
from openvino.model_zoo.model_api.adapters import OpenvinoAdapter, create_core
```

# read input image using opencv
input_data = cv2.imread("sample.png")

# define the path to mobilenet-ssd model in IR format
model_path = "public/mobilenet-ssd/FP32/mobilenet-ssd.xml"

# create adapter for OpenVINO™ runtime, pass the model path
model_adapter = OpenvinoAdapter(create_core(), model_path, device="CPU")

# create model API wrapper for SSD architecture

> ssd_model = SSD(model_adapter, preload=True)

Reviewer 4: probably it makes sense to say what preload=True means if you set it explicitly

Author: Added the comment

# apply input preprocessing, sync inference, model output postprocessing
results = ssd_model(input_data)

# visualize results and show it

> image_with_bboxes = draw_detections(input_data, results)

Reviewer 4: I do not see any imports, functions for draw_detection implementation

Author: Removed

results = ssd_model(input_data)

> ```

Reviewer 3: We won't ever verify the snippet works. To me it's the strong reason to delete it. You should refer to a demo instead

Author: If the majority prefers to delete it, I will do it

Reviewer 3: This is not about voting.
1. This is about explaining why your solution is correct.
2. Even if you insist on conducting a poll,  your voice won't count. My voice won't count either. Only the voice of the person who is responsible for OMZ matters. Which is @vladimir-dudnik's one

Author: My point is the demos contain only complex cases with asynchronous models execution, it is not the only one usage scenario. We should provide somewhere the example of simple synchronous model call.

Didn't get point about the snippet work verification. Many packages have a documentation with API examples. The package is going to be updated with new releases, and with the new releases the documentation will be also updated as well as the snippet.

Reviewer 3: You can write a dedicated sample and refer to it. You will anyway need to cover your lib with tests. The sample could be a part of tests.

> Didn't get point about the snippet work verification. Many packages have a documentation with API examples. The package is going to be updated with new releases, and with the new releases the documentation will be also updated as well as the snippet.

Such packages have people whos work to continuously check the example is consistent. We don't have such people. If packages don't do that, they usually end up having broken examples.

Reviewer 3: Here is the solution. Keep this section, but remove it after tests are added

Author: Okay

To study the complex scenarios, refer to [Open Model Zoo Python* demos](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos), where the asynchronous inference is applied.

The list of Open Model Zoo demos with Model API support:

> - BERT Named Entity Recognition Python* Demo

Reviewer 4: probably it make sense to provide link on demo readme

Author: Added links to the demos

- [BERT Question Answering Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/bert_question_answering_demo/python)
- [BERT Question Answering Embedding Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/bert_question_answering_embedding_demo/python)
- [Classification Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/classification_demo/python)
- [Image Deblurring Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/deblurring_demo/python)
- [Human Pose Estimation Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/human_pose_estimation_demo/python)
- [Instance Segmentation Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/instance_segmentation_demo/python)
- [MonoDepth Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/monodepth_demo/python)
- [Object Detection Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/object_detection_demo/python)

> - [Image Segmentation Python* Demo](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/segmentation_demo/python)

Reviewer 3: Why would we want to support another list of demos? We are already having trouble supporting root demos list. Don't add another

Author: It's suggestion from @eaidova

Reviewer 4: I just wanted to highlight where we use model api, no need additional list if we can mark demos in common list

Author: I will left this list for now. The idea was to have a small description of Model API package in OMZ docs and there refer to the Model API documentation in online doc (this description might be in a different file or in demos/readme.md). So for this purpose keeping all Python demos that support Model API might be useful.

To find such demos in a huge common demos list is not convenient

Reviewer 3: How does the list serve the purpose of having a small description of Model API. One link to an exaple demo will be enough
